File: app/src/main/java/c/m/storyapp/register/domain/use_case/user_register_use_case/UserRegisterUseCase.kt
package c.m.storyapp.register.domain.use_case.user_register_use_case

import c.m.storyapp.common.util.Resource
import c.m.storyapp.register.domain.model.RegisterResponse
import c.m.storyapp.register.domain.repository.RegisterRepository
import kotlinx.coroutines.flow.Flow

import javax.inject.Inject

class UserRegisterUseCase @Inject constructor(private val registerRepository: RegisterRepository) {
    suspend operator fun invoke(
        name: String,
        email: String,
        password: String,
    ): Flow<Resource<RegisterResponse>> {
        return registerRepository.postRegister(name, email, password)
    }
}

File: lib/register/domain/use_case/user_register_use_case.dart
import 'package:dartz/dartz.dart';
import 'package:injectable/injectable.dart';
import 'package:story_app/common/util/constants.dart';
import 'package:story_app/common/util/failure/failure.dart';
import 'package:story_app/common/util/failure/server_failure.dart';
import 'package:story_app/common/util/resource.dart';
import 'package:story_app/register/domain/model/register_response.dart';
import 'package:story_app/register/domain/repository/register_repository.dart';

@injectable
class UserRegisterUseCase {
  final RegisterRepository _registerRepository;

  UserRegisterUseCase(this._registerRepository);

  Future<Either<Failure, RegisterResponse>> call(
    String name,
    String email,
    String password,
  ) async {
    final resource = await _registerRepository.postRegister(
      name,
      email,
      password,
    );

    if (resource is Success) {
      return Right(
        resource.data ??
            const RegisterResponse(
              error: false,
              message: Constants.blankString,
            ),
      );
    } else {
      return Left(ServerFailure(message: resource.message));
    }
  }
}
